### DISCUSSION TITLE
Typing information associated with `BaseModel` `validator`

### DISCUSSION THREAD

**Discussion by NathanFarmer** [pydantic/pydantic#5050] [Question] [answered]
Is it possible to use a `validator` with `pre=True` and have `mpy` pick up on the types accepted/handled? Is there a better way of writing this code so I don't have `# type: ignore[arg-type]` in the method?

```python
from pydantic import BaseModel, validator
from typing import List, Optional


class Foo(BaseModel):
    bar: List

    @validator("bar", pre=True)
    def _validate_bar(cls, v: Optional[List]) -> List:
        if v:
            ...  # validation logic for each item
        else:
            v = []
        return v


# set mutable default to None
def get_foo(bar: Optional[list] = None) -> Foo:
    return Foo(bar=bar)  # mypy error: Argument "bar" to "Foo" has incompatible type
                         # "Optional[List[Any]]"; expected "List[Any]"  [arg-type]
```

**Comment by dmontagu** [comment, marked as answer]
This is not currently possible, and in general I don't think we are likely to make modifications so that mypy uses information about the validators to modify the signature of the `__init__` (mostly because I think it would be difficult to handle all cases correctly, and an on-going maintenance burden). I would be inclined to accept a PR for it if there was significant interest but I think it could be challenging.

In general when you are writing python code and therefore can use the "python API" for instantiating your models (as opposed to parsing JSON/etc.), my recommendation (largely for type-checker compatibility) would be to handle type conversion in your own code. E.g.:
```python
def get_foo(bar: Optional[list] = None) -> Foo:
    return Foo(bar=bar or [])
```

**Comment by NathanFarmer** [comment]
@dmontagu I appreciate the quick response! You convinced me that the answer is to forget using a `validator` for the type conversion, and handle it outside of the `BaseModel`.

The solution above won't work quite the same for me due to type conversion that gets applied in the `validator`. This demonstrates how types are changed in the `validator`, yet the `or []` solution won't work.

```
class Foo(BaseModel):
    bar: List[int]

    @validator("bar", pre=True)
    def _validate_bar(cls, v: Optional[List]) -> List:
        if v:
            ...  # transformation logic for float types
        else:
            v = []
        return v


# set mutable default to None
def get_foo(bar: Optional[list[int | float]] = None) -> Foo:
    return Foo(bar=bar or [])  # Argument "bar" to "Foo" has incompatible type 
                               # "List[Union[int, float]]"; expected "List[int]"  [arg-type]
```

One possible solution that I don't love is to call the `validator` directly:

```
class Foo(BaseModel):
    bar: List[int]

    @validator("bar", pre=True)
    def parse_bar(cls, bar: Optional[List]) -> List:
        if bar:
            ...  # transformation logic for float types
        else:
            bar = []
        return bar


def get_foo(bar: Optional[list[int | float]] = None) -> Foo:
    parsed_bar = Foo.parse_bar(bar=bar)
    return Foo(bar=parsed_bar)
```

Instead I opted to do as was recommended which was to handle type conversion outside of the `validator` like this:
```
class Foo(BaseModel):
    bar: List[int]


def parse_bar(bar: Optional[List[int | float]] = None) -> List[int]:
    if bar:
        ... # transformation logic for float types
    else:
        bar = []
    return bar


def get_foo(bar: Optional[List[int | float]] = None) -> Foo:
    parsed_bar: List[int] = parse_bar(bar=bar)
    return Foo(bar=parsed_bar)
```

**Reply by dmontagu** [reply]
For what it's worth, you can keep the `parse_bar` validator on Foo and have it just call the `parse_bar` function.

Also, I understand where you are coming from with frustration at the type system not being smarter here. I think if we did want to improve the support for a differently-type-hinted `__init__`, we'd probably set up the mypy plugin to read `Annotation`s and have an `Annotation` that indicated to the plugin how to override the `__init__` type hint. Like I said though, I suspect this may be a decent chunk of work. (Significantly less than trying to infer automatically from validators though.)

**Reply by dmontagu** [reply]
Also note that there is still the business of getting good support in IDEs (PyCharm / VSCode at least). That may be another good argument in favor of not adding mypy plugin support, since that will either require more IDE integration work or lead to a bad user experience in IDEs.

Perhaps something better might be to go down the path of something like `InitVar` (how it works with dataclasses), but I suspect that won't work perfectly with pydantic since generally the `InitVar` would have a different name than the fields in the `dataclasses` world. I haven't thought about it too much though.

**Comment by NathanFarmer** [comment]
Related: https://github.com/pydantic/pydantic/discussions/3997
